[IV-24-25] Objetivo 6

juanbarearojo opened the pull request:
* [x] ¿Se han establecido criterios *a priori* para elegir el sistema de
integración continua más conveniente y se han calificado los sistemas que
cumplan esos requisitos según esos criterios?
* [x] ¿Se han examinado varios sistemas de integración continua?
* [x] ¿Se han configurado varios sistemas de integración continua?
* [x] ¿Se ha configurado correctamente el *Checks API* en los sistemas en los que
   sea necesario para que aparezca correctamente en GitHub (y se pueda comprobar
   desde los tests)?
* [x] ¿Uno de los sistemas configurados permite comprobar cuales son las versiones
   dellenguaje con las que funciona correctamente nuestra aplicación?
* [x] ¿Se escogen de forma adecuada las versiones del lenguaje que se testean,
   tanto en el sistema de CI como en el contenedor Docker?
* [x] ¿Se han justificado correctamente las versiones del lenguaje que se están
testeando y se ha comprobado que no se comprueban varias veces lo mismo (la
misma versión en CI y en el contenedor Docker en otro sistema CI?

juanbarearojo committed:
#39 Elección criterios

Se eligen los criterios de antemano para la futura elección del sistema de integración continua. Se descarta el criterios de compatibilidad con Docker y github porque se considera obvia ya que la gran mayoria de herramientas lo incluten actualmente. closes #39

juanbarearojo committed:
#42 Prueba circle Ci

No se ha trabajado nunca con esta herramienta se sube el config ya que es necesario antes de configurar la pipeline

juanbarearojo committed:
#42 Correción typo

Se había puesto mal el nombre de la imagen. Corregido

juanbarearojo committed:
#42 Modificacion config

Se cambia el config ya que no estaba haciendo checkout y por ende estaba dando fallo al ejecutarlo. No pasaba el test

juanbarearojo committed:
#39 Travis añadido a la documentación

juanbarearojo committed:
#39 Circleci añadido

En este caso se añade circleci como una opción viable ya que respesta el hecho de que no es necesario pagar

juanbarearojo committed:
#42 Semaphore prueba.

juanbarearojo committed:
#42 Se establcer docker como agent

juanbarearojo committed:
#42 Semaphore prueba 2

juanbarearojo committed:
#42 Semaphore prueba 3

juanbarearojo committed:
#42 Semaphore prueba 4

juanbarearojo committed:
#42 Semaphore prueba 5

juanbarearojo committed:
#42 Semaphore prueba 6

juanbarearojo committed:
42 Prueba semaforo 7

juanbarearojo committed:
#42 Semaphore prueba 8

juanbarearojo committed:
#42 Semaphore prueba 9

juanbarearojo committed:
#39 Semaphore ci añadido

En caso se añade semaphore ci como una opción trás varias pruebas. Semaphore pipeline solo acepta la version v1.0 aunque actualmente exista la dos. En caso la función de este Ci es la ejecución de pruebas de los test.

juanbarearojo committed:
#39 Añadido 2 criterios nuevos

En este caso se añaden dos criterios nuevos trás probar harness. Esto debido a que ahora trás probar varias herramientas se ven necesidades que hasta este momento no se habían considerado porque no se tenía el suficiente conocimeinto.

juanbarearojo committed:
#39 Añadido 3 opciones nuevas

Se añaden tres opciones nuevas siendo buddy jenkins y harness. Buddy no fue probado ya que no genera una archivo. Jenkins no se consiguió que funcionara debido a que dió problemas el jdk. Harness se configuró entero pero no se pudo hacer commit ya que al no tner una iv. Harness no permite la ejecución de su pipeline. Aunque no se pudieron usar ninguna se quiso dejar como reflejo del trabajo realizado.

juanbarearojo committed:
#41 app Veyor probando versiones Go

En este caso se prueba el sistema appveyor. Se hace para que pruebe distintas versiones de Go. Las versiones de go probadas son las 1.23, 1,21,1,22. Esto como se explica en la https://go.dev/doc/devel/release#policy solo son mantenidas las dos ultimas versiones. Debido a esto se prueban esas dos y la anterior para ver si se podría usar una versión que ya no es mantenida. closes #41

juanbarearojo committed:
#39 appveyor añadido como una opción

App Veyor es añadido a la documentación

juanbarearojo committed:
#42 Prueba test github actiosn

Se prueba github actions con un check que prueba la imagen docker creada en el objetivo anterior

juanbarearojo committed:
#42 Modificacion DockerFile

como el usaurio no tiene permiso para escribir en la carpeta /__w/_temp/_runner_file_commands/. Hay dos opciones para solucionarlo: 1 darle permisos a test parq eu pueda modificar esa carpeta, ejecutarlo como root. Se opta por la segunda opción

juanbarearojo committed:
#42 Restaurado DockerFile

Debido a que no soluciona el problema el cambio se restaura y se ejecutara como root el actions original

juanbarearojo committed:
#42 Restaurado DockerFile 2

Se había restaurado a una versión anterior a la utilizada. Se cambia para eliminar cualquier cambio realizado sobre el dockerfile

juanbarearojo committed:
#42 Modificación workflow

Se sabe que no es la más recomendado ejecutar la imagen mediante root pero la otra posible solución es alinear el uid de github actions con el del contenedor. Esto debido a que no tiene permisos para ejecutar al hacer el checkout para escribir en la carpeta /__w/_temp/_runner_file_commands/. No se considera ue la opción de alienar los permisos y el uid sea mejor

juanbarearojo committed:
#42 Añadido action y elección

Se eligen las herramientas que se van a usar después de haberlas probado. Se mantienene el resto como muestra del trabajo realizado y buscado. Se añade actiosns a la lsita de opciones

juanbarearojo committed:
#42 Se clave Ci a iv.yaml

hossam1522 reviewed (commented):
Así por encima comentar que en la descripción del objetivo viene:

>Aprovechamiento del contenedor de Docker generado en el hito anterior en alguno de los sistemas de CI, especialmente si hay un cambio o adaptación del mismo.

La idea es usar el contenedor docker en alguno de los sistemas, no en todos. En el caso de no usar Docker, deberás buscarte alguna forma de instalar Go y Task y hacer las pruebas correspondientes prescindiendo del contenedor (consejo: en github actions es bastante sencillo)

Luego en un rato te reviso lo demás.

juanbarearojo commented:
> Así por encima comentar que en la descripción del objetivo viene:
> 
> > Aprovechamiento del contenedor de Docker generado en el hito anterior en alguno de los sistemas de CI, especialmente si hay un cambio o adaptación del mismo.
> 
> La idea es usar el contenedor docker en alguno de los sistemas, no en todos. En el caso de no usar Docker, deberás buscarte alguna forma de instalar Go y Task y hacer las pruebas correspondientes prescindiendo del contenedor (consejo: en github actions es bastante sencillo)
> 
> Luego en un rato te reviso lo demás.

Entonces voy echarle un vistazo porque llevo bastante rato dándome con esta pared la verdad y la solución  a la que había llegado no me gustaba la verdad

juanbarearojo committed:
#42 Modiicacion actions

Se modifica el workflow de github para no usar la imagen docker. Se instalan de forma manual go y task.Se elimina el problema de ejecutar como root

juanbarearojo committed:
#42 modificacion descrpicion actions

hossam1522 commented on .semaphore/semaphore.yml:
```diff
@@ -0,0 +1,14 @@
+        - name: run-tests
+          commands:
+            - checkout
+            - 'docker run --rm -v `pwd`:/app/test juanbarearojo0/privatechef'
```
Realmente no tiene mucho sentido ejecutar una imagen dentro de otra imagen.

juanbarearojo replied:
El problema lo conozco. Semaphore  pone en su documentación de su pipeline _The version of the pipeline YAML specification to be used. The only value supported is v1.0_. El problema de usar la version 1.0 es que no permite la definición no es posible declarar tu propia imagen de Docker como el entorno de ejecución . La configuración del agent en esa versión se limita a elegir uno de los tipos de máquina y un sistema operativo preestablecidos que te ofrece Semaphore

juanbarearojo replied:
Corregido 197676ac4549fb35aed4fc27bc50f03f70f1fad7

hossam1522 commented on documentation/ci.md:
```diff
@@ -0,0 +1,33 @@
+## Elección
+
+- Se elige **CircleCI** para que ejecute los tests a través de la imagen Docker.
+- Se elige **AppVeyor** para probar las distintas versiones de Go. Como se establece en este documento (https://go.dev/doc/devel/release#policy), las versiones mantenidas son las dos últimas versiones que son 1.23 y 1.22. Debido a esto, son probadas. También se prueba la versión 1.21 para comprobar si funcionaría una versión que ya no está mantenida.
```
¿Dónde se está probando esto?

![image](https://github.com/user-attachments/assets/4bf10f56-fbb6-4da4-94e2-f1da8988c468)

La idea de esto

> Para que sean detectados correctamente por GitHub (y por los tests de la asignatura), habrá que configurar el sistema para que [use el checks API](https://docs.github.com/en/rest/reference/checks).

Era precisamente que se pudiera ver.

juanbarearojo replied:
Como el último cambio fue en al documentación no aparece porque está puesto para que si el readme o la documentación se actualizan no se ejecute el test

juanbarearojo replied:
Voy a cambiarlo porque si no se puede dar loq ue acaba de pasar

juanbarearojo committed:
#42 Modificación appveyor

Debido a que estaba puesto que cuando la documentación fuera modificada no ejecutara el test se daba la situación de que parecía que no estaba bien conectado con la check api. Se cambia esto para que no se vea reflejado este fallo

juanbarearojo committed:
#42 modificacion semaphore

Es cambiado el sempahore para evitar la ejecución de una imagen dentro de otra. Ahora se intala go y task para ejecutar los test

juanbarearojo committed:
#42 Modifiación semaphore

Se cambia el type ya que estaba dando error al ser incompatible.

juanbarearojo commented:
@hossam1522 cuando puedas otra vez revisión porfavor

hossam1522 commented on iv.yaml:
```diff
@@ -6,4 +6,5 @@ automatizar:
 
-test: internal/models/asignacion_test.go
\ No newline at end of file
+test: internal/models/asignacion_test.
```
typo

juanbarearojo replied:
corregido en a31c0d55052f302d8b643bd05262e574f35cd5af

hossam1522 commented on documentation/ci.md:
```diff
@@ -0,0 +1,33 @@
+## Elección
+
+- Se elige **CircleCI** para que ejecute los tests a través de la imagen Docker.
+- Se elige **AppVeyor** para probar las distintas versiones de Go. Como se establece en este documento (https://go.dev/doc/devel/release#policy), las versiones mantenidas son las dos últimas versiones que son 1.23 y 1.22. Debido a esto, son probadas. También se prueba la versión 1.21 para comprobar si funcionaría una versión que ya no está mantenida.
```
Hmm no sé, creo recordar que si especificas 1.23 por ejemplo, te coge por defecto la 1.23.0 en vez de la 1.23.4, que sí es la última versión estable mantendida. Igual pasa con la 1.22 y la 1.22.10. (Aunque realmente no encuentro fuentes que afirmen esto).

juanbarearojo replied:
Yo había supuesto que era la revés que cuando ejecutabas se instalaba la última versión de parche disponible de esa serie. Pero mirando la salida del test appveyor se ejecuta con 1.23.2 auqneu este puesto que lo haga con 1.22
![image](https://github.com/user-attachments/assets/efb12b20-c6bc-4d84-8fb4-6ddfe6777eed)Investigando he visto que desde Go 1.21, con “GOTOOLCHAIN=auto”  se actualiza automáticamente la versión de Go si mi archivo go.mod tiene una versión  superior. Por ejemplo, si mi go.mod pone Go 1.23.2 y yo tengo instalada la versión 1.22, Go descargará automáticamente la versión 1.23.2 en tiempo de compilación Si uso “GOTOOLCHAIN=local”, la herramienta no se auto-actualiza. Sin embargo, si mi go.mod tiene una versión mayor que la instalada, el proceso fallará. Como quiero probar varias versiones de Go (como 1.21, 1.22 y 1.23), puedo establecer “go 1.21” en el archivo go.mod y configurar “GOTOOLCHAIN=local”. De esa manera, me aseguro de que mi pipeline respete la versión de Go que tengo instalada localmente y evito que se descargue otra toolchain. Voy a probar si esto soluciona el problema

juanbarearojo replied:
Al final resulta que con poner 123 se descarga la última versión 
![image](https://github.com/user-attachments/assets/f399adc8-a660-4204-9969-64e0c4c0d1b3)
![image](https://github.com/user-attachments/assets/4828152f-2cdd-4b5a-beaa-a73732f312c3)
Pero no me hubiera dado cuenta del otro fallo si no me lo hubiera dicho así que muchas gracias.

Corregido en c8e6bacbd22cd155b7991ecf68b98a14ccf73073

hossam1522 reviewed (commented):
Te dejo 2 tonterías para que les des una vuelta y diría que ya estaría.

juanbarearojo committed:
#39 correción typo iv.yaml

juanbarearojo committed:
#40 go.mod rebajado version go

Desde Go 1.21 se introdujo GOTOOLCHAIN=auto, que hace que Go descargue
una versión superior si go.mod especifica una mayor que la instalada.

Al rebajar 'go 1.23' a 'go 1.21', podemos poner GOTOOLCHAIN=local para evitar que se actualice y podemos compilar localmente con Go 1.21, 1.22 o 1.23 sin conflictos ni errores por “se requiere versión X, pero tenemos Y. Si no hacemos este cambio daría error

juanbarearojo committed:
#40 Añadido gotoolchain =local

Se añade la variable de entorno GOTOOLCHAIN=local para que Go no
se auto-actualice cuando detecta una versión superior en go.mod.

Con esto, cada versión que instalamos en la matriz de AppVeyor
(1.21, 1.22, 1.23) se usa directamente sin que Go baje otra toolchain,
permitiéndonos probar varias versiones reales sin descargas adicionales.

juanbarearojo commented:
@hossam1522 Ahora sí creo que está bien. Cuando sea posible los revisa. Muchas gracias de antemano por tu tiempo de verdad. El trabajo que estás haciendo con las revisiones no es poco la verdad.

hossam1522 reviewed (approved):
Pues, por mi parte, lo considero aprobado. ¡Buen trabajo!

juanbarearojo commented:
@JJ Cuando sea posible revisión porfavor. Al igual que con hossam gracias de antemano por tu tiempo

antoniorr02 reviewed (approved):
Yo por mi parte tambien veo bastante bien el objetivo, has encontrado incluso herramientas que yo no encontré en el mio y has hecho la elección probandolas con lo que considero que es muy buen trabajo. ¡Enhorabuena!

JJ reviewed (approved):
Lo voy a dar por bueno porque has sido exhaustivo, pero creo que estás repitiendo tests de versiones de Go; tendría que mirar qué versiones usas en Docker, sin embargo, así que voy a imaginar que tienes la versión última de desarrollo o si no que la vas a poner antes de fusionar.

juanbarearojo committed:
40 Modificación versiones probadas

Se actualiza el appveyor para evitar esta repitición ya que en el docker ya se usa la versión 1.23 por lo que es repititvo volver a probarla aquí

juanbarearojo committed:
#40 Actualización versión DockerFile

Se actualiza a la versión 1.23.4 que la estable y más reciente
